File: python/day6.py
#!/usr/local/bin/python3

with open('../day6_input.txt') as f:
    lines = [line.rsplit(maxsplit=3) for line in f.read().splitlines()]

# Part one
# lights = [[False for i in range(1000)] for j in range(1000)]

# Part two
lights = [[0 for i in range(1000)] for j in range(1000)]

for instruction, start, _, end in lines:
    start_x, start_y = [int(part) for part in start.split(',')]
    end_x, end_y = [int(part) for part in end.split(',')]

    if instruction == 'turn on':
        # func = lambda state: True         # Part one
        func = lambda state: state + 1      # Part two
    elif instruction == 'turn off':
        # func = lambda state: False        # Part one
        func = lambda state: state if state == 0 else state - 1  # Part two
    elif instruction == 'toggle':
        # func = lambda state: not state    # Part one
        func = lambda state: state + 2      # Part two

    for col in range(start_x, end_x + 1):
        row_part = slice(start_y, end_y + 1)
        lights[col][row_part] = map(func, lights[col][row_part])


print(sum(sum(col) for col in lights))

File: 2015/javascript/day6.js
#!/usr/local/bin/node

'use strict';
var fs = require('fs');

var instructions = fs.readFileSync('../day6_input.txt', 'utf-8')
    .trimRight()
    .split('\n')
    .map(line => line.match(/([\w ]+) (\d+,\d+) through (\d+,\d+)/))
    .map(match => [match[1],
                   match[2].split(',').map(num => parseInt(num, 10)),
                   match[3].split(',').map(num => parseInt(num, 10))]);
                   // The anonymous function feeding parseInt is
                   // necessary otherwise you end up with NaN.
                   // No, I don't know why.

// var grid = Array(1000).fill(Array(1000).fill(false));
// Can't do this. Array.fill reuses the same array reference,
// so changes to one affect all the filled arrays.
// This is like [[False] * 5] * 5 in Python
var grid = [];
for (var x = 0; x < 1000; x += 1) {
    grid[x] = [];
    for (var y = 0; y < 1000; y += 1) {
        // grid[x].push(false);     // Part one
        grid[x].push(0);
    }
}

var transformations = {
    // 'turn on': state => true,    // Part one
    'turn on': state => state + 1,  // Part two
    // 'turn off': state => false,  // Part one
    'turn off': state => state === 0 ? 0 : state -1,    // Part two
    // 'toggle': state => !state    // Part one
    'toggle': state => state + 2    // Part two
};

instructions.forEach(function (line) {
    var func = transformations[line[0]];
    var start_x = line[1][0];
    var start_y = line[1][1];
    var end_x = line[2][0];
    var end_y = line[2][1];

    for (var col = start_x; col < end_x + 1; col += 1) {
        for (var row = start_y; row < end_y + 1; row += 1) {
            grid[col][row] = func(grid[col][row]);
        }
    }
});

console.log([].concat.apply([], grid).reduce((prev, cur) => prev + cur))
